### PR TITLE
Enhances Dockerfile to reducing image size and use official ruby image

### PR DESCRIPTION
**Pull request by sinoru**
This PR will enhances Dockerfile to reducing image size and uses official ruby image.

### main (2f7c3cb6)
Build Time: 127.7s
Image Size: 1.37GB

### PR
Build Time: 202.1s
Image Size: 1.23GB (-0.14GB, 11.38% Reduced)

Both runs on MacBook Pro 16" M1 Max (Docker: 8 Cores, 8GB RAM)

### CODE REVIEW

**Comment by shleeable** [comment]
Welcome to the Dockerfile thunderdome! This PR might not get accepted because we're got like 5 other PRs doing similar things.

Thank you for the effort tho.

**Comment by sinoru** [comment]
> Welcome to the Dockerfile thunderdome! This PR might not get accepted because we're got like 6 other PRs doing similar things.
> 
> Thank you for the effort tho.

Ooops...
Anyway I can rebase or drop this PR for the decision that this or other PR accepted.

File: Dockerfile [hunk 1]
```diff
@@ -1,98 +1,190 @@
...
-        python3 \
-        shared-mime-info && \
-    bundle config set --local deployment 'true' && \
-    bundle config set --local without 'development test' && \
-    bundle config set silence_root_warning true && \
-    bundle install -j"$(nproc)" && \
-    yarn install --pure-lockfile --network-timeout 600000
-
-FROM node:${NODE_VERSION}
+# https://github.com/nodejs/docker-node/blob/main/16/bullseye-slim/Dockerfile
```
Comment (nschonni) [inline Dockerfile:6]: Not sure I follow the benifit of copy/pasting the official Node image text and taking on the burden of keeping it in sync, even if we don't change the template often at all

Comment (sinoru) [reply Dockerfile:6]: @nschonni Thanks to review 😀

I tried to use official image instructions for keeping image light and using specific version of nodejs, but yeah that can be problem.
But, As I know, official didn't change that template either, they just changes version args to release new version.

But if this is not a prefer way, than I might use nvm or other node version manager to install specific version easily.

Comment (sinoru) [reply Dockerfile]: @nschonni I change to use [NodeSource's binaries](https://github.com/nodesource/distributions).

**Comment by vmstan** [comment]
I just ran through and did a build with this Dockerfile and it compiled successfully, and it also looks a lot cleaner.

**Comment by sinoru** [comment]
> I would like to see if it could be flipped to use bookworm as the Debian image now that it's out, I tried but got errors related to `LoadError: libidn.so.12: cannot open shared object file: No such file or directory - /usr/local/bundle/gems/idn-ruby-0.1.5/lib/idn.so` that I haven't resolved yet. I have a bookworm version of the current Dockerfile in `main` so I know it can be done, not sure what dependency it's missing yet

@vmstan I tried.. probably `idn-ruby` and bookworm's `libidn` has some conflicts? Might be fix or workaround available, but I don't know right now...
Might be I can try later, but now in my time zone, it is almost 5 am which I needs to sleep :)

**Comment by vmstan** [comment]
When I went to deploy based off this image, it didn't work.

**Comment by sinoru** [comment]
@vmstan Now it will works.

**Comment by sinoru** [comment]
@vmstan Now it should works.

**Comment by vmstan** [comment]
Now running on vmst.io, so far so good.

**Comment by vmstan** [comment]
How can we confirm that jemalloc is working? I see where `LD_PRELOAD="libjemalloc.so.2"` is set but I've also read where that needs a full path `LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libjemalloc.so.2"` -- either way when I run something like `ruby -r rbconfig -e "puts RbConfig::CONFIG['LIBS']"` or `MAINLIBS` I don't see it listed.

I also tried switching from libjemalloc2 to libjemalloc-dev on the apt install, but didn't seem to make any difference. Would it even compile without this working?

**Comment by sinoru** [comment]
@vmstan 
> How can we confirm that jemalloc is working?

It can be checked through `/proc/$pid/maps`? I think.
```sh
➜  mastodon git:(docker) docker run -it --rm sinoru/mastodon:docker sh
$ ps -ef
UID        PID  PPID  C STIME TTY          TIME CMD
mastodon     1     0  0 15:28 pts/0    00:00:00 /usr/bin/tini -- sh
mastodon     7     1  0 15:28 pts/0    00:00:00 sh
mastodon     8     7 25 15:28 pts/0    00:00:00 ps -ef
$ cat /proc/7/maps
ffffadf30000-ffffadfb9000 r-xp 00000000 fe:01 3552544                    /usr/lib/aarch64-linux-gnu/libjemalloc.so.2
```

> Would it even compile without this working?

It should works. It likes hooking the binary.
But, might be I can update the usage on Dockerfile more better to follow [official guide](https://github.com/jemalloc/jemalloc/wiki/Getting-Started).

**Comment by vmstan** [comment]
So far I haven't seen anything to indicate that it isn't working, I just wanted to check. Sidekiq tasks don't show any abnormal memory usage but I haven't had them running very long (they don't tend to in my environment, as I adopt new commits to `main` on a less than daily basis.)

**Comment by vmstan** [comment]
kubectl exec -it mastodon-86f59f446-c2gnr -- sh

```
$ ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
mastodon       1       0  0 16:01 ?        00:00:00 /usr/bin/tini -- bundle exec puma -C config/puma.rb
mastodon       7       1  0 16:01 ?        00:00:08 puma 6.3.0 (tcp://0.0.0.0:3000) [mastodon]
mastodon      10       7  9 16:01 ?        00:04:25 puma: cluster worker 0: 7 [mastodon]
mastodon      14       7  9 16:01 ?        00:04:24 puma: cluster worker 1: 7 [mastodon]
mastodon      16       7 10 16:01 ?        00:04:36 puma: cluster worker 2: 7 [mastodon]
mastodon     364       0  0 16:46 pts/0    00:00:00 sh
mastodon     371     364 25 16:46 pts/0    00:00:00 ps -ef
$ cat /proc/7/maps | grep jem
7f5907b09000-7f5907b12000 r--p 00000000 fe:01 406859                     /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
7f5907b12000-7f5907bb8000 r-xp 00009000 fe:01 406859                     /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
7f5907bb8000-7f5907bd0000 r--p 000af000 fe:01 406859                     /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
7f5907bd0000-7f5907bd6000 r--p 000c6000 fe:01 406859                     /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
7f5907bd6000-7f5907bd7000 rw-p 000cc000 fe:01 406859                     /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
```

Looks like it's working here.

**Comment by renchap** [comment]
Thanks for your work on this!

The state of the Dockerfile is quite bad and every attempt to improve it is good. There are multiple PRs trying to improve things, but I definitely like basing off the official image.

https://github.com/mastodon/mastodon/pull/25808 has a lot of improvements regarding caching, so I will need to find some time to review both of your PRs and see how they can work together.

**Comment by sinoru** [comment]
@renchap Thanks for responding!

I can help always for resolving conflicts or rebasing this PR for working together.

Furthermore in current circumstances, I know that there are a lots of PR for improving is waiting and it will be hard to solve all conflicts at all, if #25808 or other authors agrees, than I may try or help create new PR to cherry picking those improvements it can to make apply at once with conflict free.

**Comment by vmstan** [comment]
Can you update this to reflect the removal of MASTODON_VERSION_FLAGS and MASTODON_VERSION_SUFFIX flags in favor of MASTODON_VERSION_PRERELEASE and MASTODON_VERSION_METADATA?

**Comment by sinoru** [comment]
@vmstan I merged `main` to this PR :)

**Comment by vmstan** [comment]
Cool. Once 4.2 has finalized I think this should get some traction. Been working very well for me.

**Comment by sinoru** [comment]
This PR is replaced by #26861

**Pull request closed without merging**